wal-vdom/src/virtual_dom/velement.rs

```diff
@@ -1,45 +1,70 @@
-use std::collections::HashMap;
-use gloo::{utils::document, console::__macro::JsValue};
+use gloo::{console::log, utils::document};
 use serde::Serialize;
-use web_sys::Node;
+use std::collections::HashMap;
+use web_sys::Element;
 
 use super::VNode;
 
-
-#[derive(Serialize)]
+#[derive(Debug, Serialize)]
 pub struct VElement {
     pub tag_name: String,
     pub attr: HashMap<String, String>,
     pub children: Vec<VNode>,
 }
 
 impl VElement {
-    pub fn new(
-        tag_name: String,
-        attr: HashMap<String, String>,
-        children: Vec<VNode>,
-    ) -> VElement {
+    // TODO: maybe some types for attributes and children
+    // List of attributes - https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes
+    //                    - https://www.w3schools.com/tags/ref_attributes.asp
+    // Maybe some oop approach with defined types for attributes and types for elements?
+    pub fn new(tag_name: String, attr: HashMap<String, String>, children: Vec<VNode>) -> VElement {
         VElement {
             tag_name,
             attr,
             children,
         }
     }
 
-    pub fn render(&self) -> Result<Node, JsValue> {
-        let el = document()
-            .create_element(self.tag_name.as_str())?;
+    /// Renders virtual Element into concrete DOM Element object. Diffs on tag name,
+    /// attributes and children
+    pub fn render(&self, target: &mut Element, last: Option<&VElement>) {
+        match last {
+            Some(last) if last.tag_name == self.tag_name => {
+                log!("\t\tComparing attrs");
+                // Compare attributes
+                for (key, val) in self.attr.iter() {
+                    target
+                        .set_attribute(key, val)
+                        .expect("Couldnt set attribute");
+                }
+                for (key, _val) in last.attr.iter() {
+                    if !self.attr.contains_key(key) {
+                        target
+                            .remove_attribute(key)
+                            .expect("Couldnt remove attribute");
+                    }
+                }
+            }
+            _ => {
+                // inverted check, if last == None || last = Some(x) that x.tag_name !=
+                // self.tag_name => Swap whole element
+                log!("\t\tRendering new node");
+                let new_el = document()
+                    .create_element(&self.tag_name)
+                    .expect("Couldnt create new element");
 
-        // set attributes
-        for (key, val) in self.attr.iter() {
-            el.set_attribute(key, val)?;
-        }
-        // appending children
-        for v_child in self.children.iter() {
-            let child = v_child.render()?;
-            el.append_child(&child)?;
-        }
+                // add attributes
```

Comment:
zamiast komentarza "add attributes" imo można by było poprostu wydzielić to do funkcji. Tak samo w paru innych miejscach żeby nie było jak w yew funkcji koszmarków 500 linijkowych